README.md

```diff
@@ -14,15 +14,106 @@ gem 'lance'
 
 And then execute:
 
-    $ bundle install
+```bash
+$ bundle install
+```
 
 Or install it yourself as:
 
-    $ gem install lance
+```bash
+$ gem install lance
+```
 
 ## Usage
 
-TODO: Write usage instructions here
+### Draft
+
+```ruby
+class Log
+  include ::Lance::Something
+end
+
+log = ::Lance::Log.build(STDOUT)
+log.renderer = ::Lance::Renderer.build
+log.filter = ::Lance::Filter.build
+
+msg = log.build_message
+
+msg.info("Foo", tags: [:some, :tags])
+
+log.tags(:foo).info("message")
+log.info(["message", field1: 123], tags: [:foo])
+log.tags(:foo).info("message", foo: 123, bar: "ooo")
+```
+
+#### Syntax
+
+```ruby
+log = ::Lance::Log.build(STDOUT)
+
+# Alternative 1
+log.tags(:foo, :bar).info("message")
+log.tags(:foo, :bar).info("message", foo: 123, bar: "ooo")
+log.tags(:foo, :bar).info("message", anything_that_can_become_json)
+log.tags(:foo, :bar).info(anything_that_can_become_json) do
+  "dynamic parsed message"
+end
+log.tags(:foo, :bar).info(foo: 123, bar: "ooo") do
+  "dynamic parsed message"
+end
+log.info("without tags")
+log = log.tags(:foo, :bar)
+log.tags(:baz).info("message") # 3 tags: :foo, :bar, :baz
+
+# Alternative 2
```

Comment:
I like this second alternative better :+1: but all of them look good! What I liked most was the list of tags and calling `log.info` directly, for example.

Reply:
Hey @stefannibrasil !
I'm a bit conflicted on the second one, while it's my favourite in term of tagging, it makes the the "json data" approach a bit hostile.
It mixes "reserved things" and "non-reserved things".

I'm thinking mainly of two use-cases:
1. Small amount of data, provided manually by the developer
2. Big json data structure, provided as a whole

Based on my experience, the first case is a lot more common:

```ruby
log.info("foo", tags: [:one, :two], id: id, type: "Something", total: amount)
```

The output looking like

```
[2020-01-01T15:15:00.000000Z] Foo::Bar#method {one, two} INFO: foo {"id": 1, "type": "Something", total: 100}
```

While the other case is more specialized for big data structures, for example let's pretend you have a user object and you want to log all data related to the user:

```ruby
log.info("foo", tags: [:one, :two], data: decision)
```

In this case, I'd expect the content of `data` to go straight to a new line:

```
[2020-01-01T15:15:00.000000Z] Foo::Bar#method {one, two} INFO: foo
{
  "id": 1,
  "last_ip": "192.168.1.1",
  "encrypted_password": "someweirdtext",
  "email": "blah"
  ...
}
```

This implies the following things:
- Any newly added key is backward incompatible, if I decide to introduce a key `:inspect` to render objects as you call `inspect` on them, any usage of `:inspect` key for the first case will break, essentially requiring an update
- Alternatively, you could force any other key to always be inside `:data`, but I find this hostile to usage

I can't think of a solution for that, but I think the first case is quite important and should be handled. I can see an alternative where I dedicate a totally different method to logging data, for example:

```ruby
log.data(foo).info("message before data")
# or if no message
log.data(foo).info
# or if no message but with tags
log.data(foo).info(tags: [:data])
```

But the first one is the one that's been bothering me and I don't know how to solve.
Any suggestion? :laughing:

Reply:
I just had another interesting and challenging thought: 
How do I handle messages with invalid data? Should it raise an exception?

Check the following example:

```ruby
@ ~ welcome $ pry
[1] pry(main)> def foo(tags:, **args)
[1] pry(main)*   puts tags.inspect
[1] pry(main)*   puts args.inspect
[1] pry(main)* end  
=> :foo
[3] pry(main)> tmp={tags: 555, asd: "loool"}
=> {:tags=>555, :asd=>"loool"}
[4] pry(main)> foo(tags: [:a_tag], tmp: "mu", **tmp)
555
{:tmp=>"mu", :asd=>"loool"}
=> nil
[5] pry(main)> 
```

So calling `**tmp` overwrites tags, let's presume the developer didn't notice it, but tags is also _not a symbol array_ which is what I expect, instead.

I have two alternatives: I can print an error message from the logger itself, or raise an exception.
I'm a bit conflicted about raising an exception because a logger is used everywhere, although it seems correct.
At the same time, an error message might be overlooked.